### PR TITLE
Feature/aks cluster creation

### PR DESCRIPTION
**Pull request by rpudlowski93**
PR with AKS cluster setup using Terraform scripts. Connected with task #1440 
There is a small description of the code in readme and I will appreciate if someone makes a review and give me some feedback :)

### CODE REVIEW

File: feature/aks-cluster-creation/azure-subnet.tf [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# DO NOT Modify by hand - Managed by Automation
+#####################################################
+#####################################################
+# This file can be used as a base template to build other Terraform files. It attempts to use as much
+# Terraform interprolation as possible by creating Terraform variables instead of changing inline
+# this approach provides an easier way to do creative looping, fetch IDs of created resources etc.
+#####################################################
+#####################################################
+# {{ specification.name }}
+#####################################################
```
Comment (mkyc) [inline feature/aks-cluster-creation/azure-subnet.tf:11]: how is that managed by automation?

File: feature/aks-cluster-creation/azure-vnet.tf [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#####################################################
+# DO NOT Modify by hand - Managed by Automation
+#####################################################
+#####################################################
+# This file can be used as a base template to build other Terraform files. It attempts to use as much
+# Terraform interprolation as possible by creating Terraform variables instead of changing inline
+# this approach provides an easier way to do creative looping, fetch IDs of created resources etc.
+#####################################################
+#####################################################
+#####################################################
```
Comment (mkyc) [inline feature/aks-cluster-creation/azure-vnet.tf:10]: how is that managed by automation?

File: feature/aks-cluster-creation/main.tf [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+# Azure Resource Manager provider
+# https://www.terraform.io/docs/providers/azurerm/index.html
+provider "azurerm" {
+  version = ">= 2.15.0"
+  features {}
```
Comment (mkyc) [inline feature/aks-cluster-creation/main.tf:6]: do we need this?

File: feature/aks-cluster-creation/vars.tf [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  description = "Resource id of existing the Virtual Network subnet where You want to join Your AKS cluster"
+  default = "/subscriptions/2d60775f-932a-4cf6-b9f0-548a8b43b368/resourceGroups/aks-cluster-test/providers/Microsoft.Network/virtualNetworks/aks-cluster-network/subnets/aks-subnet"
+}
+
+variable "existing_resource_group_name" {
+  description = "Name of the existing AKS cluster resource group where You want to join Your AKS cluster"
+  default = "aks-cluster-test"
+}
+
+terraform { required_version = ">= 0.12" }
```
Comment (mkyc) [inline feature/aks-cluster-creation/vars.tf:115]: move it to main

File: feature/aks-cluster-creation/vars.tf [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    min_count                      = null
+    max_count                      = null
+    }
+}
+
+
+########### EXISTING NETWORK
+variable "existing_vnet_subnet_id" {
+  description = "Resource id of existing the Virtual Network subnet where You want to join Your AKS cluster"
+  default = "/subscriptions/2d60775f-932a-4cf6-b9f0-548a8b43b368/resourceGroups/aks-cluster-test/providers/Microsoft.Network/virtualNetworks/aks-cluster-network/subnets/aks-subnet"
```
Comment (mkyc) [inline feature/aks-cluster-creation/vars.tf:107]: omg

File: feature/aks-cluster-creation/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+## Terraform scripts to provision AKS cluster #
+***
+The Azure Kubernetes Service (AKS) is the AWS service for deploying, managing, and scaling containerized applications with Kubernetes. This repo contain terraform files, to automate the AKS cluster deployment.
```
Comment (mkyc) [inline feature/aks-cluster-creation/README.md:3]: > is the AWS service

oh, really?

File: feature/aks-cluster-creation/azure-subnet.tf (hunk 1 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/azure-subnet.tf]: Removed, historical lines, I don't know why but I didn't focused on these lines/comments.

File: feature/aks-cluster-creation/azure-vnet.tf (hunk 2 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/azure-vnet.tf]: Removed, historical lines, I don't know why but I didn't focused on these lines/comments.

File: feature/aks-cluster-creation/vars.tf (hunk 4 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/vars.tf]: Done, thanks!

File: feature/aks-cluster-creation/README.md (hunk 6 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/README.md]: Long day.. Tired :D Fixed

File: feature/aks-cluster-creation/main.tf (hunk 3 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/main.tf:4]: The "feature" block is required for AzureRM provider 2.x.

**Comment by rpudlowski93** [comment]
> First question is why is [that module](https://registry.terraform.io/modules/Azure/aks/azurerm/4.0.0) not utilised here? Some other comments also added.

I took into account the module but in my opinion, after research and opinions a few people the modules for AKS are not enough "developed". AKS module which You metioned, require service principle which should be created before but better option here is to use Managed System Identity which is something like wrapper to service principle, is "passwordless", do not have expiration date and the corresponding service principal is automatically removed when managed identity is removed. Managed System Identity is created when the resource is created and it is automatically removed when the resource is deleted. Moreover, the module does not have optional inputs such as: network plugin, public ip, different os, autoscaling and much more what we would like to have in Epiphany. 

Summing up, mentioned module use azurerm_resource_group and I decided to use directly the resource without the module in order to better adapt AKS cluster setup to our needs.

File: feature/aks-cluster-creation/vars.tf [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+variable "aks_address_prefix" {
+  description = "SubNetwork address space"
+  default = ["10.1.1.0/24"]
+}
+
+variable "public_ssh_key_path" {
+    description = "Public ssh key path"
+    default     = "/workspaces/epiphany/core/src/epicli/ssh/id_rsa.pub"
```
Comment (mkyc) [inline feature/aks-cluster-creation/vars.tf:43]: should be something more like `~/.ssh/some_public_rsa.pub`

File: feature/aks-cluster-creation/azure-resource-groups.tf [hunk 8]
```diff
@@ -0,0 +1,5 @@
+resource "azurerm_resource_group" aks-rg {
```
Comment (mkyc) [inline feature/aks-cluster-creation/azure-resource-groups.tf:1]: missing optional quote

File: feature/aks-cluster-creation/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
+## Terraform scripts to provision AKS cluster #
+***
+The Azure Kubernetes Service (AKS) is the service for deploying, managing, and scaling containerized applications with Kubernetes. This repo contain terraform files, to automate the AKS cluster deployment.
+***
+#### Nice to know:
+- AKS require two resources group. First resource group which is managed by user and contains only the Kubernetes service resource. Second resource group, known as the node resource group, contains all of the infrastructure resources associated with the cluster. These resources include the Kubernetes node VMs, virtual networking, and storage. AKS automatically remove the node resource group whenever the cluster is removed.
+- AKS require service principle to create additional resources but managed identities are essentially a wrapper around service principals, and make their management simpler. It is better to use Managed identities since are passwordless.
```
Comment (mkyc) [inline feature/aks-cluster-creation/README.md:7]: typo: principle

File: feature/aks-cluster-creation/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
+## Terraform scripts to provision AKS cluster #
+***
+The Azure Kubernetes Service (AKS) is the service for deploying, managing, and scaling containerized applications with Kubernetes. This repo contain terraform files, to automate the AKS cluster deployment.
+***
+#### Nice to know:
+- AKS require two resources group. First resource group which is managed by user and contains only the Kubernetes service resource. Second resource group, known as the node resource group, contains all of the infrastructure resources associated with the cluster. These resources include the Kubernetes node VMs, virtual networking, and storage. AKS automatically remove the node resource group whenever the cluster is removed.
```
Comment (mkyc) [inline feature/aks-cluster-creation/README.md:6]: I do not remember exactly how, but I was able to force AKS via terraform to do it all in one RG. Let's move it to separate task.

File: feature/aks-cluster-creation/main.tf (hunk 3 above)
Comment (mkyc) [reply feature/aks-cluster-creation/main.tf:4]: ok.

File: feature/aks-cluster-creation/README.md (hunk 10 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/README.md:6]: Hymm.. intresting, I will take a look but I was sure that it is not possible (based on azure documentation): 

https://docs.microsoft.com/en-us/azure/aks/faq:
```
As you work with the node resource group, keep in mind that you **cannot**:

Specify an **existing** resource group for the node resource group.
```

File: feature/aks-cluster-creation/README.md (hunk 9 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/README.md:7]: Sure, thanks, fixed :)

File: feature/aks-cluster-creation/azure-resource-groups.tf (hunk 8 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/azure-resource-groups.tf:1]: Thanks! :)

File: feature/aks-cluster-creation/vars.tf (hunk 7 above)
Comment (rpudlowski93) [reply feature/aks-cluster-creation/vars.tf:43]: True :) Fixed

**Comment by rpudlowski93** [comment]
> some little comments and newlines everywhere ;)

Thanks! :) New lines added :)

File: feature/aks-cluster-creation/README.md (hunk 10 above)
Comment (mkyc) [reply feature/aks-cluster-creation/README.md]: Checked once again, and in fact I was wrong. AKS creates separate RG for it's resources.

File: feature/aks-cluster-creation/vars.tf [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    node_count                     = 0
+    vm_size                        = "Standard_DS2_v2"
+    os_type                        = "Windows"
+    os_disk_size_gb                = "50"
+    type                           = "VirtualMachineScaleSets"
+    enable_auto_scaling            = false
+    min_count                      = null
+    max_count                      = null
+    }
+}
```
Comment (pprach) [inline feature/aks-cluster-creation/vars.tf:116]: Just a suggestion maybe it will make sense to use this:
https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files

Comment (mkyc) [reply feature/aks-cluster-creation/vars.tf:116]: Those are good point. That being said, my opinion is that we should add another followup tasks for those 2 things @pprach mentioned and we should close this task and move to autoscaling.

Comment (rpudlowski93) [reply feature/aks-cluster-creation/vars.tf]: Yes, that's true! I was thinking of it but I wasn't sure if it is necessary. It is a easy and fast change which I've just done.. Could You take a look @mkyc and @pprach at last commit and approve and close the PR if everything is correct? 

Regarding to creating own module, I will discuss the topic with @toszo and create new ticket for it.

File: feature/aks-cluster-creation/terraform.tfvars [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    type                           = "VirtualMachineScaleSets"
+    enable_auto_scaling            = false
+    min_count                      = null
+    max_count                      = null
+    }
+additional_cluster_node_pools = {
+    name                           = "windows"
+    node_count                     = 0
+    vm_size                        = "Standard_DS2_v2"
+    os_type                        = "Windows"
```
Comment (pprach) [inline feature/aks-cluster-creation/terraform.tfvars:27]: why windows in defaults?

Comment (rpudlowski93) [reply feature/aks-cluster-creation/terraform.tfvars:27]: This is additional node pool with 0 node count (because we would like to have different options). The default node pool has by default Linux since AKS need linux os :)
 Summing up: 
- Default node pool - Linux 
- Additional node pool - Windows/Linux

Comment (pprach) [reply feature/aks-cluster-creation/terraform.tfvars:27]: Yes I got it but still think it should be linux as default. But approved.
